test/spectests/spectests.cpp

```diff
@@ -0,0 +1,290 @@
+#include "execute.hpp"
+#include "parser.hpp"
+#include <nlohmann/json.hpp>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+namespace fs = std::filesystem;
+using json = nlohmann::json;
+
+namespace
+{
+constexpr auto JsonExtension = ".json";
+
+template <typename T>
+uint64_t json_to_value(const json& v)
+{
+    return static_cast<uint64_t>(static_cast<T>(std::stoull(v.get<std::string>())));
+}
+
+class test_runner
+{
+public:
+    void run_from_file(const fs::path& path)
+    {
+        std::cout << "Running tests from " << path << "\n";
+
+        std::ifstream test_file{path};
+        json j;
+        test_file >> j;
+
+        for (const auto& cmd : j.at("commands"))
+        {
+            const auto type = cmd.at("type").get<std::string>();
+
+            std::cout << "Line " << cmd.at("line").get<int>() << ": " << type << " " << std::flush;
```

Comment:
Since there are pass/skip/fail helpers maybe it is worth considering a `log` or `info` helper so that the output stream handling is done at a single place.

Reply:
Added this, but now this requires some additional `to_string` calls